Lib/asyncio/threads.py

```diff
@@ -0,0 +1,21 @@
+"""High-level support for working with threads in asyncio"""
+
+import functools
+
+from . import events
+
+
+__all__ = "to_thread",
+
+
+async def to_thread(func, /, *args, **kwargs):
+    """Asynchronously run function *func* in a separate thread.
+
+    Any *args and **kwargs supplied for this function are directly passed
+    to *func*.
+
+    Return an asyncio.Future which represents the eventual result of *func*.
+    """
+    loop = events.get_running_loop()
+    func_call = functools.partial(func, *args, **kwargs)
```

Comment:
I think this line should be changed to:
```suggestion
    context = contextvars.copy_context()
    func_call = functools.partial(context.run, functools.partial(func, *args, **kwargs))
```

because it can take unexpected behavior with contextvars.

For example i modify you'r example with adding usage of contextvars:
```python3
import asyncio
import time
from contextvars import ContextVar

ctx_value = ContextVar("ctx_value")


def blocking_io():
    print(f"start blocking_io at {time.strftime('%X')}")
    # Note that time.sleep() can be replaced with any blocking
    # IO-bound operation, such as file operations.
    time.sleep(1)
    print(f"value from context in blocking_io {ctx_value.get()}")
    print(f"blocking_io complete at {time.strftime('%X')}")


async def main():
    print(f"started main at {time.strftime('%X')}")

    ctx_value.set(42)

    await asyncio.gather(
        asyncio.to_thread(blocking_io),
        asyncio.sleep(1))

    print(f"finished main at {time.strftime('%X')}")


asyncio.run(main())
```

and got an error:
```
started main at 07:37:48
start blocking_io at 07:37:48
Traceback (most recent call last):
  File "/opt/project/asd.py", line 30, in <module>
    asyncio.run(main())
  File "/usr/local/lib/python3.10/asyncio/runners.py", line 43, in run
    return loop.run_until_complete(main)
  File "/usr/local/lib/python3.10/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/opt/project/asd.py", line 25, in main
    await asyncio.gather(asyncio.to_thread(blocking_io), asyncio.sleep(1))
  File "/usr/local/lib/python3.10/asyncio/threads.py", line 21, in to_thread
    return await loop.run_in_executor(None, func_call)
  File "/usr/local/lib/python3.10/concurrent/futures/thread.py", line 52, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/opt/project/asd.py", line 16, in blocking_io
    print(f"value from context in blocking_io {ctx_value.get()}")
LookupError: <ContextVar name='ctx_value' at 0x7f7ba2cb7130>
```
in this case i expects correct contextual value inside thread but currently it doesn't works as well

Reply:
> in this case i expects correct contextual value inside thread but currently it doesn't works as well

I believe that is by design of `ContextVar`; in general, it is not shared across OS threads. `ContextVar` is primarily intended to be used within the same thread. For example, this also wouldn't work:

```py
from contextvars import ContextVar
import threading

ctx_value = ContextVar("ctx_value")

def worker():
    print(f"ctx_value in separate thread: {ctx_value.get()}")

def main():
    ctx_value.set(42)
    th = threading.Thread(target=worker)
    th.start()
    th.join()

if __name__ == '__main__':
    main()
```

I also think that in general that we don't want to implicitly share state across threads, as the above change would effectively do. However, we may want to consider adding a flag to `to_thread()` which allows for opt-in (`False` by default) passing of context state via something like the above. I personally don't know of the practical use cases for passing context to a separate thread though, so this will likely have to be a proposal in a separate bpo issue.

As the one who authored `contextvar` in the first place, what do you think of this @1st1?

Reply:
In this case I think we should propagate the context. That's important for tracing APIs, decimal contexts etc. The only reason why this isn't already implemented for `run_in_executor()` is because it also allowed running process pools with which contexts don't play well.

The current implementation should be amended before 3.9beta2 to essentially do this:

```python
    loop = events.get_running_loop()
    context = contextvars.copy_context()
    func_call = functools.partial(context.run, func, *args, **kwargs)
    return await loop.run_in_context(None, func_call)
```

Please submit a PR, Kyle.

Reply:
Okay, seems reasonable enough to me. I also need to add the `versionadded` to the docs anyways, I should have time later tonight to do that.

Reply:
This has been addressed in https://github.com/python/cpython/pull/20278, thanks for the feedback @JrooTJunior.